make/devices.cmake

```diff
@@ -0,0 +1,754 @@
+    list(APPEND SOURCE_FILES_DEPS ${SRC_TARGET})
+  endforeach()
+
+  set(
+    COMPLETE_YOSYS_SCRIPT
+    "${YOSYS_SCRIPT} $<SEMICOLON> write_blif -attr -cname -param ${OUT_EBLIF}"
+  )
+
+  add_custom_command(
+    OUTPUT ${OUT_EBLIF}
+    DEPENDS ${SOURCE_FILES} ${SOURCE_FILES_DEPS} ${DIRECTORY_TARGET}
+    COMMAND
+      ${YOSYS} -p "${COMPLETE_YOSYS_SCRIPT}" ${SOURCE_FILES}
+    VERBATIM
+  )
+  add_custom_target(${NAME}_eblif DEPENDS ${OUT_EBLIF})
+  add_custom_target(${NAME}_synth DEPENDS ${OUT_EBLIF})
+  add_output_to_fpga_target(${NAME} EBLIF ${OUT_LOCAL_REL}/${TOP}.eblif)
+
+  # Generate routing and generate HLC.
+  set(OUT_ROUTE ${OUT_LOCAL}/${TOP}.route)
+  set(OUT_HLC ${OUT_LOCAL}/${TOP}.hlc)
+
+  set(VPR_DEPS "")
+  list(APPEND VPR_DEPS ${OUT_EBLIF})
+
+  get_file_location(OUT_RRXML_REAL_LOCATION ${OUT_RRXML_REAL})
+  get_file_location(DEVICE_MERGED_FILE_LOCATION ${DEVICE_MERGED_FILE})
+
+  foreach(SRC ${DEVICE_MERGED_FILE} ${OUT_RRXML_REAL})
+    get_file_location(SRC_LOCATION ${SRC})
+    get_file_target(SRC_TARGET ${SRC})
+    list(APPEND VPR_DEPS ${SRC_LOCATION})
+    list(APPEND VPR_DEPS ${SRC_TARGET})
+  endforeach()
+
+  get_target_property_required(VPR env VPR)
+  set(
+    VPR_CMD
+    ${VPR}
+    ${DEVICE_MERGED_FILE_LOCATION}
+    ${OUT_EBLIF}
+    --device ${DEVICE_FULL}
+    --min_route_chan_width_hint ${VPR_ROUTE_CHAN_MINWIDTH_HINT}
+    --route_chan_width ${VPR_ROUTE_CHAN_WIDTH}
+    --read_rr_graph ${OUT_RRXML_REAL_LOCATION}
+    --verbose_sweep on
+    --allow_unrelated_clustering off
+    --max_criticality 0.0
+    --target_ext_pin_util 0.7
+    --max_router_iterations 500
+    --routing_failure_predictor off
+    --clock_modeling route
+    --constant_net_method route
+  )
+
+  # Generate IO constraints file.
+  # -------------------------------------------------------------------------
+  set(OUT_IO "")
+  if(NOT ${ADD_FPGA_TARGET_INPUT_IO_FILE} STREQUAL "")
+    get_file_location(INPUT_IO_FILE ${ADD_FPGA_TARGET_INPUT_IO_FILE})
+    get_file_target(INPUT_IO_FILE_TARGET ${ADD_FPGA_TARGET_INPUT_IO_FILE})
+    get_target_property_required(PLACE_TOOL ${ARCH} PLACE_TOOL)
+    get_target_property_required(PLACE_TOOL_CMD ${ARCH} PLACE_TOOL_CMD)
+    get_target_property_required(PINMAP_FILE ${DEVICE} ${PACKAGE}_PINMAP)
+    get_file_location(PINMAP ${PINMAP_FILE})
+    get_file_target(PINMAP_TARGET ${PINMAP_FILE})
+    set(OUT_IO ${OUT_LOCAL}/${TOP}_io.place)
+    string(CONFIGURE ${PLACE_TOOL_CMD} PLACE_TOOL_CMD_FOR_TARGET)
+    separate_arguments(
+      PLACE_TOOL_CMD_FOR_TARGET_LIST UNIX_COMMAND ${PLACE_TOOL_CMD_FOR_TARGET}
+    )
+    add_custom_command(
+      OUTPUT ${OUT_IO}
+      DEPENDS
+        ${OUT_EBLIF}
+        ${INPUT_IO_FILE}
+        ${INPUT_IO_FILE_TARGET}
+        ${PINMAP}
+        ${PINMAP_TARGET}
+        ${VPR_DEPS}
+      COMMAND ${PLACE_TOOL_CMD_FOR_TARGET_LIST} --out ${OUT_IO}
+      WORKING_DIRECTORY ${OUT_LOCAL}
+    )
+
+    set(VPR_CMD ${VPR_CMD} --fix_pins ${OUT_IO})
```

Comment:
You can't give `--fix_pins` to `--pack` IIRC?

Reply:
This was from old make system, I'll update it.